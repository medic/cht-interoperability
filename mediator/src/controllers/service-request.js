const axios = require('axios');
const {FHIR} = require('../../config');
const {url: fhirUrl, password: fhirPassword, username: fhirUsername} = FHIR;
const logger = require('../../logger');

async function createServiceRequest(request) {
  try {
    const {patientId} = request;

    const options = {
      auth: {
        username: fhirUsername,
        password: fhirPassword,
      }
    };

    const res = await axios.get(`${fhirUrl}/Patient/${patientId}`, options);

    if (res.status !== 200) {
      return {status: res.status, data: res.data};
    }

    // todo => @njogz to add task creation process.

    return {status: res.status, data: res.data};
  } catch ({response: res}) {
    logger.error(JSON.stringify(res.data, null, 4));
    return {status: res.status, data: res.data};
  }
}

module.exports = {
  createServiceRequest,
};
