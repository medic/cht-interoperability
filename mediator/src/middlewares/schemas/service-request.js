const joi = require('joi');

const createServiceSchema = joi.object({
  patientId: joi.number().required()
});

module.exports = {
  createServiceSchema,
};
