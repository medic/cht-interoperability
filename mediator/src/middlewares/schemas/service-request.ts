import joi from 'joi';

const RequesterSchema = joi.object({
  reference: joi.string().regex(/Organization\/\S+/),
});

export const ServiceRequestSchema = joi.object({
  intent: joi.string().required(),
  subject: joi
    .object({
      reference: joi.string().regex(/Patient\/\S+/),
    })
    .required(),
  requester: RequesterSchema.required(),
});
