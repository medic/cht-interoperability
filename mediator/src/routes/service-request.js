const {Router} = require('express');
const {createServiceRequest} = require('../controllers/service-request');
const {requestHandler} = require('../utils/request');
const {validateBodyAgainst} = require('../middlewares');
const {createServiceSchema} = require('../middlewares/schemas/service-request');

const router = Router();

router.post(
  '/',
  validateBodyAgainst(createServiceSchema),
  requestHandler(req => createServiceRequest(req.body))
);

module.exports = router;
