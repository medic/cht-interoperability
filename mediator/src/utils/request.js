function requestHandler(handler) {
  return  (req, res) => {
    handler(req)
    .then(({status, data}) => {
      res.status(status).send(data);
    });
  };
}

module.exports = {
  requestHandler,
};
