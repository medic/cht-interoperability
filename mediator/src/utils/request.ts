import { Request, Response } from 'express';

type RequestHandler = (req: Request) => Promise<{status: number, data: any}>;

export function requestHandler(handler: RequestHandler) {
  return  (req: Request, res: Response) => {
    handler(req)
    .then(({status, data}) => {
      res.status(status).send(data);
    });
  };
}
