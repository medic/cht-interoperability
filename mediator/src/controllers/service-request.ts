import axios from "axios";
import { logger } from "../../logger";
import { FHIR, CHT } from "../../config";
import { generateFHIRSubscriptionResource } from "../utils/subscription";
import { generateApiUrl } from "../utils/service-request";
import https from "https";

const { url: fhirUrl, password: fhirPassword, username: fhirUsername } = FHIR;

type ServiceRequest = {
  patient_id: string;
  callback_url: string;
};

async function createServiceRequest(request: ServiceRequest) {
  try {
    const { patient_id: patientId, callback_url } = request;

    let options: Record<string, any> = {
      auth: {
        username: fhirUsername,
        password: fhirPassword,
      },
    };

    const res = await axios.get(`${fhirUrl}/Patient/?identifier=${patientId}`, options);

    if (res.status !== 200) {
      return { status: res.status, data: res.data };
    }

    // generate subscription resource
    const FHIRSubscriptionResource = generateFHIRSubscriptionResource(
      patientId,
      callback_url
    );
    const subscriptionRes = await axios.post(
      `${fhirUrl}/Subscription`,
      FHIRSubscriptionResource,
      options
    );

    if (subscriptionRes.status !== 201) {
      return { status: subscriptionRes.status, data: subscriptionRes.data };
    }

    // call the CHT API to set up the follow up task
    const chtApiUrl = generateApiUrl(CHT.url, CHT.username, CHT.password);
    options = {
      httpsAgent: new https.Agent({
        rejectUnauthorized: false,
      }),
    };
    
    const body = {
      _meta: {
        form: "interop_follow_up",
      },
      patient_uuid: patientId,
    };
    
    const chtRes = await axios.post(chtApiUrl, body, options);

    if (chtRes.data.success !== true) {
      // TODO: delete the subscription
      return { status: 500, data: { message: "unable to create the follow up task"} };
    }

    logger.info(JSON.stringify(chtRes.data, null, 4));

    return { status: res.status };
  } catch (err) {
    logger.error(`Error: ${err}`);
    return { status: 500 };
  }
}

module.exports = {
  createServiceRequest,
};
