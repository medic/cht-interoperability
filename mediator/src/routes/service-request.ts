import { validateBodyAgainst } from '../middlewares';
import { requestHandler } from '../utils/request';

const { Router } = require('express');
const { createServiceRequest } = require('../controllers/service-request');
const {
  createServiceSchema,
} = require('../middlewares/schemas/service-request');

const router = Router();

router.post(
  '/',
  validateBodyAgainst(createServiceSchema),
  requestHandler((req) => createServiceRequest(req.body))
);

export default router;
