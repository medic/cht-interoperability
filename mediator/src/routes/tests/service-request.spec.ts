import { createServiceRequest } from '../../controllers/service-request';
import request from 'supertest';
import app from '../../..';
import { ServiceRequestFactory } from '../../middlewares/schemas/tests/fhir-resource-factories';

jest.mock('../../controllers/service-request');

describe('POST /service-request', () => {
  it('accepts incoming request with valid service request resource', async () => {
    (createServiceRequest as any).mockResolvedValueOnce({
      data: {},
      status: 201,
    });

    const data = ServiceRequestFactory.build();
    data.status = 'active';

    const res = await request(app).post('/service-request').send(data);

    expect(res.status).toBe(201);
    expect(res.body).toEqual({});
    expect(createServiceRequest).toHaveBeenCalledWith({
      ...data,
      resourceType: 'ServiceRequest',
    });
    expect(createServiceRequest).toHaveBeenCalled();
  });

  it('doesn\'t accept incoming request with invalid service request resource', async () => {
    const data = ServiceRequestFactory.build({ request: undefined });

    const res = await request(app).post('/service-request').send(data);

    expect(res.status).toBe(400);
    expect(res.body.message).toMatchInlineSnapshot(`""ServiceRequest.status" Missing property"`);
    expect(createServiceRequest).not.toHaveBeenCalled();
  });
});
