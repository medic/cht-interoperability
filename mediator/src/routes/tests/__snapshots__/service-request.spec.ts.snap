// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /service-request calls handler for valid incoming request 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "callback_url": "https://callback.medic.org",
        "patient_id": "PATIENT_ID",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`POST /service-request returns a bad request error for invalid incoming request 1`] = `
{
  "message": ""callback_url" must be a valid uri",
}
`;
